core/ir/aarch64/codec.c

```diff
@@ -802,7 +802,7 @@ encode_fpimm8_half(opnd_t opnd, uint abc_offset, uint defgh_offset, OUT uint *en
      *   abbb cdef gh00 0000
      */
 #ifdef HAVE_HALF_FLOAT
-#    if !defined(DR_HOST_NOT_TARGET) && !defined(STANDALONE_DECODER)
+#    if !defined(BUILD_TESTS)
```

Comment:
I'm again wondering if there should be a different trigger for something like this as it seems like it could not be what we want in some situations where we build tests

Reply:
But whether we build the test suite or not is today an independent decision that does not affect the core DR libraries: they do not become test-only libraries just because we're building the suite.  I'm also not sure that there is no real use of encoding something that isn't supported by the hardware: there could be some kind of cross-platform operation a tool wants to do.  Consider someone recording an address trace on some hardware and then post-processing it on other hardware and that post-processing wanting to record some instruction encodings into the trace file, based on the recorded hardware and not the post-processing hardware.

Reply:
> I'm also not sure that there is no real use of encoding something that isn't supported by the hardware: there could be some kind of cross-platform operation a tool wants to do.

Currently the above example is a build-time feature check, `HAVE_HALF_FLOAT`, along with a test-only build time directive, `BUILD_TESTS`. But most are runtime feature checks in the generated encoder conditional on `BUILD_TESTS`, e.g.
```
    case OP_abs:
#       if !defined(BUILD_TESTS)
        if (!proc_has_feature(FEATURE_SVE))
            return ENCFAIL;
#       endif
        return encode_opndsgen_04112000_00c01fff(pc, instr, 0x0416a000, di);
. . .
    case OP_cas:
#       if !defined(BUILD_TESTS)
        if (!proc_has_feature(FEATURE_LSE))
            return ENCFAIL;
#       endif
        enc = encode_opndsgen_08a07c00_001f03ff(pc, instr, 0x88a07c00, di);
```
Does that mean we should remove these guards as well? Currently these methods enable us to run codec tests independent of host h/w feature support. How can we have a build which does this on unsupported h/w as well as encoding correctly on supported h/w in e.g. clients and tools.

Reply:
This seems like the vector length and ISA variants that today are programmably set by the user: if the user can set whether FEATURE_SVE should be considered to be available then these tests can turn that on (irrespective of the real underlying hardware) and a user with some cross-platform encoding use case could also turn it on.  Does that seem reasonable?  I guess it would be dr_set_proc_features() or sthg like that?

Reply:
> Does that seem reasonable? I guess it would be dr_set_proc_features() or sthg like that?

Yes, that makes sense. If I understand correctly, users should have the ability to make DR "pretend" that it's running on SVE h/w or not, regardless of the h/w, in order to use the codec to encode as well as decode? Internally we can also use `dr_set_proc_features()` for encoding tests on SVE and non-SVE h/w.

I think this approach is related to the comments raised here https://github.com/DynamoRIO/dynamorio/pull/5835#discussion_r1087279135

Perhaps it's possible to unify the approach to handle both cases.

Reply:
> If I understand correctly, users should have the ability to make DR "pretend" that it's running on SVE h/w or not, regardless of the h/w, in order to use the codec to encode as well as decode? Internally we can also use `dr_set_proc_features()` for encoding tests on SVE and non-SVE h/w.

Yes, exactly.

Reply:
Resolved. See https://github.com/DynamoRIO/dynamorio/pull/5835#discussion_r1245466005.